router/cmd/main.go

```diff
@@ -42,88 +49,176 @@ func Main() {
 		os.Exit(0)
 	}
 
-	result, err := config.LoadConfig(*configPathFlag, *overrideEnvFlag)
-	if err != nil {
-		log.Fatalf("Could not load config: %s", err)
+	// We load this after flag parse so that "OVERRIDE_ENV" can be set by dotenv OR the flag
+	if *overrideEnvFlag != "" {
+		_ = godotenv.Overload(*overrideEnvFlag)
+	}
+
+	/*
+		Config path precedence:
+		1. Flag
+		2. Environment variable
+		3. Dotenv loaded environment variable
+		4. Default config file
+	*/
+
+	configPath := *configPathFlag
+
+	// If not set by flag or normal environment variable, check again for dotenv override loaded envar
+	if configPath == "" {
+		configPath = os.Getenv("CONFIG_PATH")
 	}
 
-	logLevel, err := logging.ZapLogLevelFromString(result.Config.LogLevel)
+	// This is redundant with the check inside config.LoadConfig, but we need to have the path
+	// here as well to watch it.
+	if configPath == "" {
+		configPath = config.DefaultConfigPath
+	}
+
+	result, err := config.LoadConfig(configPath)
 	if err != nil {
-		log.Fatalf("Could not parse log level: %s", err)
+		log.Fatalf("Could not load config: %s", err)
 	}
 
-	logger := logging.New(!result.Config.JSONLog, result.Config.DevelopmentMode, logLevel).
+	logLevelAtomic := zap.NewAtomicLevelAt(result.Config.LogLevel)
+
+	baseLogger := logging.New(!result.Config.JSONLog, result.Config.DevelopmentMode, logLevelAtomic).
 		With(
 			zap.String("service", "@wundergraph/router"),
 			zap.String("service_version", core.Version),
 		)
 
 	// Start pprof server if address is provided
 	if *pprofListenAddr != "" {
-		pprofSvr := profile.NewServer(*pprofListenAddr, logger)
+		pprofSvr := profile.NewServer(*pprofListenAddr, baseLogger)
 		defer pprofSvr.Close()
 		go pprofSvr.Listen()
 	}
 
 	// Start profiling if flags are set
-	profiler := profile.Start(logger, *cpuProfilePath, *memProfilePath)
+	profiler := profile.Start(baseLogger, *cpuProfilePath, *memProfilePath)
 	defer profiler.Finish()
 
-	// Handling shutdown
-	ctx, stop := signal.NotifyContext(context.Background(), os.Interrupt,
-		syscall.SIGHUP,  // process is detached from terminal
-		syscall.SIGTERM, // default for kill
-		syscall.SIGKILL,
-		syscall.SIGQUIT, // ctrl + \
-		syscall.SIGINT,  // ctrl+c
-	)
-	defer stop()
-
-	if *configPathFlag != "" {
-		logger.Info(
-			"Config file path provided. Values in the config file have higher priority than environment variables",
-			zap.String("config_file", *configPathFlag),
-		)
-	} else if result.DefaultLoaded {
-		logger.Info("Found default config file. Values in the config file have higher priority than environment variables",
-			zap.String("config_file", config.DefaultConfigPath),
+	sl := baseLogger.With(zap.String("component", "supervisor"))
+
+	rs := supervisor.NewRouterSupervisor(&supervisor.RouterSupervisorOpts{
+		Logger: sl,
+
+		LifecycleHooks: &supervisor.LifecycleHooks{
+			LoadResources: func(rr *supervisor.RouterResources) error {
```

Comment:
What is the benefit of the callback if the supervisor is already coupled to the core / router ?